augur/util_support/shell_command_runner.py

```diff
@@ -0,0 +1,90 @@
+import os
+import sys
+import subprocess
+from textwrap import dedent
+
+
+class ShellCommandRunner:
+    """
+    Run the given command string via Bash with error checking.
+
+    TODO move to method docstrings
+    Returns True if the command exits normally.  Returns False if the command
+    exits with failure and "raise_errors" is False (the default).  When
+    "raise_errors" is True, exceptions are rethrown.
+
+    If an *extra_env* mapping is passed, the provided keys and values are
+    overlayed onto the default subprocess environment.
+    """
+
+    def __init__(self, cmd, *, raise_errors=False, extra_env=None):
+        self.cmd = cmd
+        self.raise_errors = raise_errors
+        self.extra_env = extra_env
+
+    def run(self):
+        try:
+            self.invoke_command()
+        except Exception as error:
+            self.print_error_message(error)
+
+            if self.raise_errors:
+                raise error
+
+            return False
+
+        return True
+
+    def invoke_command(self):
+        return subprocess.check_output(
+            self.shell_executable + self.shell_args,
+            shell=False,
+            stderr=subprocess.STDOUT,
+            env=self.modified_env,
+        )
+
+    @property
+    def shell_executable(self):
+        if os.name == "posix":
+            return ["/bin/bash"]
+        else:
+            # We try best effort on other systems. For now that means nt/java.
+            return ["env", "bash"]
+
+    @property
+    def shell_args(self):
+        return ["-c", "set -euo pipefail; " + self.cmd]
+
+    @property
+    def modified_env(self):
+        env = os.environ.copy()
+
+        if self.extra_env:
```

Comment:
Would it make sense to use a [ChainMap](https://docs.python.org/3/library/collections.html#collections.ChainMap) for this?

Reply:
@jstoja This would be a clever improvement to the current code. I've opted to merge this PR as it is (or rather through the rebase at #577), just because it primarily refactors existing code without changing any other logic.

If you are up for adding the ChainMap improvement to this class, I would love to review it and add it to master.